### PR TITLE
Revert "Revert "TxnCoordSender's heartbeat uses Context cancellation when available""

### LINKED ISSUE #5559: timeout in sql TestLogic, txn:197
https://circleci.com/gh/cockroachdb/cockroach/15215

https://circle-artifacts.com/gh/cockroachdb/cockroach/15215/artifacts/0/tmp/circle-artifacts.EVY4hXL/test.log
Delays in this test: https://github.com/cockroachdb/cockroach/blob/master/sql/testdata/txn

We have a transaction that apparently gets stuck due to an earlier, aborted txn.
```
BEGIN TRANSACTION
UPDATE kv SET v = 'b' WHERE k in ('a')
BEGIN TRANSACTION    <-- error here as expected
...
ROLLBACK TRANSACTION
... some unrelated stuff ...

BEGIN TRANSACTION
UPDATE kv SET v = 'b' WHERE k in ('a')      <--- this gets stuck until the test times out some 40 seconds later
```

```
testdata/txn:197 root: UPDATE kv SET v = 'b' WHERE k in ('a')
I160324 22:11:13.051400 storage/replica_command.go:1086  failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
W160324 22:11:13.051738 kv/dist_sender.go:676  failed to invoke PushTxn [/Table/51/1/"a"/2/1,/Min): txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.02541374 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
I160324 22:11:13.051869 client/db.go:457  failed batch: txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.08630948 iso=SERIALIZABLE stat=PENDING epo=1 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
I160324 22:11:13.051924 storage/intent_resolver.go:99  [node=1,store=1,range=1] on Scan: txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.08630948 iso=SERIALIZABLE stat=PENDING epo=1 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
W160324 22:11:13.051983 storage/store.go:1671  conflicting intents on [/Table/51/1/"a"/2/1]: resolved? false
I160324 22:11:13.102662 storage/replica_command.go:1086  failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
W160324 22:11:13.103171 kv/dist_sender.go:676  failed to invoke PushTxn [/Table/51/1/"a"/2/1,/Min): txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.02541374 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
I160324 22:11:13.103354 client/db.go:457  failed batch: txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.03228220 iso=SERIALIZABLE stat=PENDING epo=1 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
I160324 22:11:13.103453 storage/intent_resolver.go:99  [node=1,store=1,range=1] on Scan: txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.03228220 iso=SERIALIZABLE stat=PENDING epo=1 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
W160324 22:11:13.103514 storage/store.go:1671  conflicting intents on [/Table/51/1/"a"/2/1]: resolved? false
I160324 22:11:13.210310 storage/replica_command.go:1086  failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
W160324 22:11:13.210829 kv/dist_sender.go:676  failed to invoke PushTxn [/Table/51/1/"a"/2/1,/Min): txn "sql/executor.go:380 sql txn" id=e4cfa1bb key=/Min rw=false pri=0.02541374 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.050232167,0 orig=1458857473.050232167,0 max=1458857473.050232167,0 wto=false failed to push "sql/executor.go:380 sql txn" id=c82252c9 key=/Table/51/1/"a"/2/1 rw=true pri=0.03126385 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458857473.046150991,0 orig=1458857473.046150991,0 max=1458857473.046150991,0 wto=false
```

This is the same issue noticed in https://github.com/cockroachdb/cockroach/issues/5376#issuecomment-198805820

### PR DESCRIPTION
**Pull request by andreimatei**
Fixes #5559

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5664)

### CODE REVIEW

**Comment by tamird** [comment]
nit: why is the commit message hand crafted? if you just `git revert ....`, then solve the merge conflicts and "git commit" you'll get the default (and descriptive) commit message.

**Comment by andreimatei** [comment]
done

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5664#-:-KE1t0y8AXO6_HHiRQQu:677588487)*

File: sql/session.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ import (
 	"net"
 	"time"
 
+	"golang.org/x/net/context"
+
```
Comment (tamird) [inline sql/session.go]: remove this space, idk why it's here

**Comment by tamird** [comment]
LGTM

**Comment by andreimatei** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/session.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5664#-KE1qvOR2KcxnNXnhk0m-r1-26:-KE2PIEaHKjqI4alC8c8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/801c7e299a6b8c769b4672ee6854bd671b8e00e1/sql/session.go#L26)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5664)*

**Pull request merged**